### PR TITLE
refactor: simplify context handling further

### PR DESCRIPTION
**Pull request by rgrinberg**
Make the context abstract

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

File: src/dune_rules/context.mli [hunk 1]
```diff
@@ -41,59 +41,59 @@ module Env_nodes : sig
...
+val env_nodes : t -> Env_nodes.t
+val instrument_with : t -> Lib_name.t list
+val profile : t -> Profile.t
+val merlin : t -> bool
 val which : t -> string -> Path.t option Memo.t
 val equal : t -> t -> bool
 val hash : t -> int
 val to_dyn : t -> Dyn.t
 val to_dyn_concise : t -> Dyn.t
+val path : t -> Path.t list
```
Comment (Alizter) [inline src/dune_rules/context.mli:73]: Can you add a comment saying that this is meant to represent the `PATH` variable.

File: src/dune_rules/context.mli [hunk 2]
```diff
@@ -41,59 +41,59 @@ module Env_nodes : sig
...
+val dynamically_linked_foreign_archives : t -> bool
+
+val ocaml : t -> Ocaml_toolchain.t
+val build_context : t -> Build_context.t
+val kind : t -> Kind.t
+val findlib_paths : t -> Path.t list
+val installed_env : t -> Env.t
+val default_ocamlpath : t -> Path.t list
+val findlib_toolchain : t -> Context_name.t option
+val env_nodes : t -> Env_nodes.t
```
Comment (Alizter) [inline src/dune_rules/context.mli:64]: This is missing a comment from
```
  ; build_dir : Path.Build.t (** env node that this context was initialized with *)
```
which was formatted incorrectly before.

File: src/dune_rules/context.mli [hunk 3]
```diff
@@ -41,59 +41,59 @@ module Env_nodes : sig
...
-  ; path : Path.t list (** [PATH] *)
-  ; ocaml : Ocaml_toolchain.t
-  ; installed_env : Env.t
-  ; findlib_paths : Path.t list
-  ; findlib_toolchain : Context_name.t option (** Misc *)
-  ; default_ocamlpath : Path.t list
-  ; build_context : Build_context.t
-  ; instrument_with : Lib_name.t list
-  }
+type t
```
Comment (Alizter) [inline src/dune_rules/context.mli:44]: A little more comments would be nice, but doesn't have to be in this PR.

File: src/dune_rules/artifact_substitution.ml [hunk 4]
```diff
@@ -94,11 +94,14 @@ module Conf = struct
...
 
   let sign_hook_of_context (context : Context.t) =
-    let config = context.ocaml.ocaml_config in
+    let config = (Context.ocaml context).ocaml_config in
     match Ocaml_config.system config, Ocaml_config.architecture config with
     | "macosx", "arm64" ->
       let codesign_name = "codesign" in
-      (match Bin.which ~path:context.path codesign_name with
+      (match
+         let path = Context.path context in
```
Comment (Alizter) [inline src/dune_rules/artifact_substitution.ml:102]: Why not just inline this?

File: src/dune_rules/context.ml [hunk 5]
```diff
@@ -80,33 +80,52 @@ module Env_nodes = struct
   ;;
 end
 
-type t =
-  { name : Context_name.t
-  ; kind : Kind.t
-  ; profile : Profile.t
+type shared =
```
Comment (Alizter) [inline src/dune_rules/context.ml:83]: So this is just currying some shared arguments in the later functions?

File: src/dune_rules/install_rules.ml [hunk 6]
```diff
@@ -730,11 +738,13 @@ end = struct
                       (Dune_package.Entry.Deprecated_library_name
                          { loc; old_public_name; new_public_name }))
             in
-            let sections = sections ctx.name [] pkg in
+            let context_name = Context.name ctx in
```
Comment (Alizter) [inline src/dune_rules/install_rules.ml:741]: This can be inlined.

**Review by Alizter — APPROVED** [review]

File: src/dune_rules/context.mli (hunk 3 above)
Comment (rgrinberg) [reply src/dune_rules/context.mli:44]: Added some.

File: src/dune_rules/artifact_substitution.ml (hunk 4 above)
Comment (rgrinberg) [reply src/dune_rules/artifact_substitution.ml:102]: No reason. It's a huge PR and I don't have time to optimize every little instance.

File: src/dune_rules/context.ml (hunk 5 above)
Comment (rgrinberg) [reply src/dune_rules/context.ml:83]: Yes, it just saves us passing a bunch of arguments in a couple of places.

**Review by Alizter — APPROVED** [review]

**Pull request closed without merging**